### PR TITLE
feat: 44 sklearn compatible interface (with fkiraly)

### PR DESCRIPTION
**Pull request by astrogilda**
Implements #44

---

Update by @fkiraly:

* merges configs and classes for public interface - configs are still used internally, but not exposed to hte user
* makes all `kwargs` parameters explicit
* completes docstrings with a description of all valid kwargs

### CODE REVIEW

**Comment by fkiraly** [comment]
Tests are failing because you *always* need to import the config classes in `base_bootstrap.py`, but you only import them for type checking. (the imports are under an "if" condition)

Are you running tests locally?

**Comment by astrogilda** [comment]
Huh I can't remember when or why I put those impirts under an if condition.
I can remove them and recommit.

Nah this time I did not run them locally, relied on github actions to test
them.

**Comment by fkiraly** [comment]
well, they tested them alright

anyway, more tests are passing now, and the properties no longer seem a problem!

So we are now properly running the tests!
Some findings:

* `BaseTimeSeriesBootstrap` should `super.__init__` (and every class with an `__init__` should)

File: src/tsbootstrap/base_bootstrap.py [hunk 1]
```diff
@@ -37,10 +48,28 @@ class BaseTimeSeriesBootstrap(BaseObject):
...
+        n_bootstraps : Integral, default=10
+            The number of bootstrap samples to create.
+        rng : Integral or np.random.Generator, default=np.random.default_rng()
+            The random number generator or seed used to generate the bootstrap samples.
+        """
+        self.n_bootstraps = n_bootstraps
+        self.rng = rng
+        self.config = BaseTimeSeriesBootstrapConfig(
+            n_bootstraps=n_bootstraps, rng=rng
+        )
```
Comment (fkiraly) [inline src/tsbootstrap/base_bootstrap.py:70]: super init missing

**Comment by fkiraly** [comment]
question: do we want to allow `__init__` `kwargs`? If yes, we have to override the standard tests. It is safer to remove `kwargs` in class constructors, but it will probably not cause too many problems if we leave them.

**Comment by fkiraly** [comment]
overriding `test_constructor` by inheritance in `TestAllBootstraps`

**Comment by astrogilda** [comment]
Okay, I can take a look.

**Comment by fkiraly** [comment]
> Okay, I can take a look.

I'd say, let's keep with the plan for now?

**Comment by astrogilda** [comment]
Wait, what plan? Sorry I'm confused. I thought you said if one wants to
retain kwargs then the tests need to be changed?

**Comment by fkiraly** [comment]
> I thought you said if one wants to
retain kwargs then the tests need to be changed?

Yes, but that wasn't your ticket - so to avoid clashing PR, I'm happy to make the change (if we decide so), and you continue working on the API change?

**Comment by astrogilda** [comment]
aaaah. Yes, please go ahead. I won't touch that issue then.

Best,
Sankalp Gilda, Ph.D.

File: src/tsbootstrap/base_bootstrap.py (hunk 1 above)
Comment (astrogilda) [reply src/tsbootstrap/base_bootstrap.py:70]: added

**Comment by fkiraly** [comment]
really strange sporadic errors on specific os-version combinations. The association with the os-version combination seems stable.

I would suspect this an actual error rather than one introduced in the interface rework, but of course it could be the latter to.

**Comment by fkiraly** [comment]
MRE for the failure:

```python
import numpy as np
from tsbootstrap.block_bootstrap import BaseBlockBootstrap

bootstrap_type = "stationary"
n_bootstraps=1
rng=None
X=np.array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

block_length = np.random.randint(1, int(0.8 * X.shape[0]) - 1)
bootstrap = BaseBlockBootstrap(
    bootstrap_type=bootstrap_type,
    block_length=block_length,
    n_bootstraps=n_bootstraps,
    rng=rng,
)

list(bootstrap.bootstrap(X))
```

**Comment by fkiraly** [comment]
Is this maybe a masked bug on `main`? The code on `main` does not seem to enter `BlockGenerator`,  while the code on this branch does.

**Pull request merged**